### PR TITLE
Move validating conventions to the model validator

### PR DESCRIPTION
**Pull request by ajcvickers**
Issue #4016

Also make sure validation logging is sent to the validation category, which means that `DiagnosticsLoggers` can be removed.

There are some convention that do work and can throw exceptions. These have not been moved.

### CODE REVIEW

File: src/EFCore/Extensions/ModelExtensions.cs [hunk 1]
```diff
@@ -17,6 +17,20 @@ namespace Microsoft.EntityFrameworkCore
...
+        ///     Returns a value indicating whether the entity types matching the given name should be configured
+        ///     as owned types when discovered.
+        /// </summary>
+        /// <param name="model"> The model to get the value from. </param>
+        /// <param name="clrType"> The type of the entity type that might be owned. </param>
+        /// <returns>
+        ///     <c>true</c> if a matching entity type should be configured as owned when discovered,
+        ///     <c>false</c> otherwise.
+        /// </returns>
+        public static bool ShouldBeOwned([NotNull] this IModel model, [NotNull] Type clrType)
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/ModelExtensions.cs]: This shouldn't be defined on `IModel`, it only makes sense on `IMutableModel` or `IConventionModel`

File: src/EFCore/Metadata/IEntityType.cs [hunk 2]
```diff
@@ -16,6 +16,12 @@ public interface IEntityType : ITypeBase
         /// </summary>
         IEntityType BaseType { get; }
 
+        /// <summary>
+        ///     Gets or sets a value indicating whether the entity type has no keys.
+        ///     If set to <c>true</c> it will only be usable for queries.
+        /// </summary>
+        bool IsKeyless { get; }
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/IEntityType.cs]: Redundant, `FindPrimaryKey() == null`

File: src/EFCore/Metadata/Internal/EntityTypeExtensions.cs [hunk 3]
```diff
@@ -28,6 +28,51 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Internal
...
+        public static IReadOnlyList<string> GetIgnoredMembers([NotNull] this IEntityType entityType)
+            => entityType.AsEntityType().GetIgnoredMembers();
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public static bool IsIgnored(
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/EntityTypeExtensions.cs]: This and `GetIgnoredMembers` shouldn't be defined on `IEntityType`, they make no sense for a readonly model.

File: src/EFCore/Metadata/Internal/ModelExtensions.cs [hunk 4]
```diff
@@ -17,6 +18,51 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Internal
...
+            ConfigurationSource configurationSource)
+            => model.IsIgnored(new TypeIdentity(name), configurationSource);
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public static bool IsIgnored(
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/ModelExtensions.cs]: Same for these

**Comment by ajcvickers** [comment]
@AndriySvyryd General question. Currently model validation runs on `IModel` (as opposed to `IMutableModel` or `IConventionModel`.) Given that those methods in your comments above should not be on the read-only types, what is the appropriate thing to do in the model validator?

**Comment by AndriySvyryd** [comment]
@ajcvickers It runs on `IModel` to indicate that it won't alter the model in any way, but its purpose is to run only on mutable models, since we assume that readonly models were already validated. So I think it's ok to downcast if needed.

**Comment by AndriySvyryd** [comment]
Strictly speaking we should introduce another readonly interface that provides access to all information from a mutable model, but I think that's overkill.

**Comment by ajcvickers** [comment]
@AndriySvyryd Sounds good

File: src/EFCore/Metadata/IConventionTypeBase.cs [hunk 5]
```diff
@@ -36,6 +36,13 @@ public interface IConventionTypeBase : ITypeBase, IConventionAnnotatable
         /// <param name="name"> The name of the member to be removed. </param>
         void RemoveIgnored([NotNull] string name);
 
+        /// <summary>
+        ///     Indicates whether the given member name is ignored.
+        /// </summary>
+        /// <param name="name"> The name of the member that might be ignored. </param>
+        /// <returns> <c>true</c> if the given member name is ignored. </returns>
+        bool IsIgnored([NotNull] string name);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/IConventionTypeBase.cs]: This is not needed. Calling `FindIgnoredConfigurationSource` would be enough

File: src/EFCore/Extensions/ConventionModelExtensions.cs [hunk 6]
```diff
@@ -41,6 +41,16 @@ public static IConventionEntityType FindEntityType([NotNull] this IConventionMod
             [NotNull] IConventionEntityType definingEntityType)
             => (IConventionEntityType)((IModel)model).FindEntityType(type, definingNavigationName, definingEntityType);
 
+        /// <summary>
+        ///     Indicates whether the given entity type name is ignored.
+        /// </summary>
+        /// <param name="model"> The model to check for ignored type. </param>
+        /// <param name="type"> The entity type that might be ignored. </param>
+        /// <returns> <c>true</c> if the given entity type name is ignored. </returns>
+        public static bool IsIgnored([NotNull] this IConventionModel model, [NotNull] Type type)
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/ConventionModelExtensions.cs]: Not needed. Can just call `GetIsIgnoredConfigurationSource`.

Though it should be renamed to `FindIgnoredConfigurationSource`

File: test/EFCore.Relational.Tests/RelationalModelValidatorTest.cs [hunk 7]
```diff
@@ -101,10 +109,19 @@ public virtual void Detects_alternate_key_with_default_value()
...
+            AddProperties(entityA);
+
             var entityB = model.AddEntityType(typeof(B));
             SetPrimaryKey(entityB);
+            AddProperties(entityB);
+            entityB.AddIgnored(nameof(B.A));
+            entityB.AddIgnored(nameof(B.AnotherA));
+            entityB.AddIgnored(nameof(B.ManyAs));
+
+
```
Comment (AndriySvyryd) [inline test/EFCore.Relational.Tests/RelationalModelValidatorTest.cs]: extra line

**Pull request merged**